### PR TITLE
Binding for Folding@home

### PR DESCRIPTION
**Pull request by fa2k**
Here is a binding for Folding@home (http://folding.stanford.edu/). It supports connecting to multiple Folding clients (computers) using the TCP interface. It can start and pause folding, and does polling of the status.

If you're asking yourself "why?", I use it successfully to augment the heating in my flat (since we're using resistive heating anyway, and a it's bit small to invest in something more efficient).

[This was double posted with my other account, sorry about that. I closed the other pull request]

### CODE REVIEW

File: addons/binding/org.openhab.binding.folding/ESH-INF/binding/binding.xml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<binding:binding id="folding"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xmlns:binding="http://eclipse.org/smarthome/schemas/binding/v1.0.0"
+        xsi:schemaLocation="http://eclipse.org/smarthome/schemas/binding/v1.0.0 http://eclipse.org/smarthome/schemas/binding-1.0.0.xsd">
+
+    <name>Folding@home</name>
+    <description>Control Folding@home clients on computers.</description>
+    <author>FaMaKe</author>
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/ESH-INF/binding/binding.xml]: could you either add your real name here or completely remove the author element?

File: addons/binding/org.openhab.binding.folding/ESH-INF/i18n/folding_xx_XX.properties [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# binding
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/ESH-INF/i18n/folding_xx_XX.properties]: please remove this file

File: addons/binding/org.openhab.binding.folding/ESH-INF/thing/thing-types.xml [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        <channels>
+            <channel id="run" typeId="run"/>
+            <channel id="finish" typeId="finish"/>
+        </channels>
+            
+        <config-description>
+            <parameter name="host" type="text">
+                <label>Host</label>
+                <description>Hostname or IP address</description>
+                <required>true</required>
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/ESH-INF/thing/thing-types.xml]: remove (as you define a default)

File: addons/binding/org.openhab.binding.folding/ESH-INF/thing/thing-types.xml [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                <default>127.0.0.1</default>
+            </parameter>
+            <parameter name="port" type="integer">
+                <label>Port</label>
+                <description></description>
+                <required>true</required>
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/ESH-INF/thing/thing-types.xml]: remove (as you define a default)

File: addons/binding/org.openhab.binding.folding/ESH-INF/thing/thing-types.xml [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                <default>36330</default>
+            </parameter>
+            <parameter name="password" type="text">
+                <label>Password</label>
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/ESH-INF/thing/thing-types.xml]: please add
```
<context>password</context>
```
to make UIs mask this entry.

File: addons/binding/org.openhab.binding.folding/ESH-INF/thing/thing-types.xml [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            <channel id="run" typeId="run"/>
+            <channel id="finish" typeId="finish"/>
+            <channel id="status" typeId="status"/>
+        </channels>
+        
+        <config-description>
+            <parameter name="id" type="text">
+                <label>Slot ID</label>
+                <description>Slot ID, "00", "01", etc.</description>
+                <required>true</required>
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/ESH-INF/thing/thing-types.xml]: remove

File: addons/binding/org.openhab.binding.folding/ESH-INF/thing/thing-types.xml [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+                <default>00</default>
+            </parameter>
+        </config-description>
+    </thing-type>
+    
+    <channel-type id="description">
+        <item-type>String</item-type>
+        <label>Description</label>
+        <description>Description of the item type.</description>
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/ESH-INF/thing/thing-types.xml]: Which "item type"? Can you provide some better description for "description"...?

File: addons/binding/org.openhab.binding.folding/META-INF/MANIFEST.MF [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+ org.eclipse.smarthome.core.library.types,
+ org.eclipse.smarthome.core.thing,
+ org.eclipse.smarthome.core.thing.binding,
+ org.eclipse.smarthome.core.thing.binding.builder,
+ org.eclipse.smarthome.core.thing.type,
+ org.eclipse.smarthome.core.types,
+ org.openhab.binding.folding,
+ org.openhab.binding.folding.handler,
+ org.osgi.framework,
+ org.osgi.service.component.annotations;version="1.2.0",
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/META-INF/MANIFEST.MF]: remove, you do not seem to use them

File: addons/binding/org.openhab.binding.folding/README.md [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+The Folding@home TCP interface is enabled only on localhost by default, without
+a password. In order to allow control of Folding on other machines than the one
+running OpenHAB, it is necessary to configure the Folding client to accept commands
+from a non-localhost address. Here is how to do it in the FAHControl application:
+
+  * Open FAHControl on the client to be added
+  * Click on Configure, then the Remote Access tab
+  * Enter a password twice (invent one)
+  * Locate the Allow box under IP Address Restrictions
+  * Append a space and the IP address of the machine running OpenHAB to the text
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/README.md]: typo: should be openHAB

File: addons/binding/org.openhab.binding.folding/README.md [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+Finish. It polls for the status of the client, updates the run / pause state, and
+provides a basic description of the slot. 
+
+The clients must be added manually in the Paper UI, but the slots for that 
+client will then appear using auto-discovery.
+
+## Requirements (network access to F@H)
+The Folding@home TCP interface is enabled only on localhost by default, without
+a password. In order to allow control of Folding on other machines than the one
+running OpenHAB, it is necessary to configure the Folding client to accept commands
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/README.md]: typo: should be openHAB

File: addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/FoldingBindingConstants.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+ */
+package org.openhab.binding.folding;
+
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+
+/**
+ * The {@link FoldingBinding} class defines common constants, which are
+ * used across the whole binding.
+ *
+ * @author FaMaKe - Initial contribution
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/FoldingBindingConstants.java]: please put your real name here

File: addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/discovery/FoldingDiscoveryProxy.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
+package org.openhab.binding.folding.discovery;
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/discovery/FoldingDiscoveryProxy.java]: please re-run "mvn license:format" to add correct license headers.

File: addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/handler/FoldingClientHandler.java [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+ * sent to one of the channels.
+ *
+ * @author FaMaKe - Initial contribution
+ */
+public class FoldingClientHandler extends BaseBridgeHandler {
+
+    private Logger logger = LoggerFactory.getLogger(FoldingClientHandler.class);
+
+    private ScheduledFuture<?> refreshJob;
+    private final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/handler/FoldingClientHandler.java]: remove - the BaseBridgeHandler already provides a protected field "scheduler".

File: addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/handler/FoldingClientHandler.java [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+            e.printStackTrace();
+            disconnected();
+        }
+    }
+
+    @Override
+    public void initialize() {
+        BigDecimal period = (BigDecimal) getThing().getConfiguration().get("polling");
+        if (period != null && period.longValue() != 0) {
+            refreshJob = scheduler.scheduleAtFixedRate(new Runnable() {
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/handler/FoldingClientHandler.java]: better use withFixedDelay

File: addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/handler/FoldingClientHandler.java [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+            s.getOutputStream().write(("slot-info\r\n").getBytes());
+            socketReader.readLine(); // Discard PyON header
+            JsonReader jr = new JsonReader(socketReader);
+            jr.setLenient(true);
+            Type slotListType = new TypeToken<List<SlotInfo>>() {
+            }.getType();
+
+            slotList = gson.fromJson(jr, slotListType);
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/handler/FoldingClientHandler.java]: remove and add a debug logging instead

File: addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/handler/FoldingClientHandler.java [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+        }
+        boolean running = false, finishing = true;
+        for (SlotInfo si : slotList) {
+            finishing &= "FINISHING".equals(si.status);
+            running |= "FINISHING".equals(si.status) || "RUNNING".equals(si.status);
+            SlotUpdateListener listener = slotUpdateListeners.get(si.id);
+            if (listener != null) {
+                listener.refreshed(si);
+            } else {
+                logger.debug("Providing a new discovery result for slot " + si.id);
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/handler/FoldingClientHandler.java]: use parameterized logging

**Comment by fa2k** [comment]
Thanks a lot for the review! We will definitely get those changes done, but not sure about the timeline.

**Comment by fa2k** [comment]
I'm having some problems running Maven. It doesn't work on Ubuntu 16.10, but there's a lot of problems with Java on that platform. I'm also having trouble in a CentOS 7 container, getting an error:

    $ mvn license:format
    org.apache.maven.InternalErrorException: Internal error: java.lang.RuntimeException: Could not instantiate required component
    ...
    Caused by: org.codehaus.plexus.component.repository.exception.ComponentLookupException: com.google.inject.ProvisionException: Guice provision errors:
    ...
    Reason: java.lang.RuntimeException: java.lang.IllegalStateException: Service is not registered class='interface org.eclipse.tycho.core.shared.ProxyServiceFacade'
      while locating org.eclipse.tycho.core.DependencyResolver annotated with @com.google.inject.name.Named(value=p2)

Full error: http://www.fa2k.net/misc/maven-error.txt

Do you have an idea about how to resolve this?

**Comment by kaikreuzer** [comment]
Sorry, I have never seen such an error...
Are you using OpenJDK? If so, it might be worth to test it with OracleJDK.

**Comment by fa2k** [comment]
The maven command worked fine with Oracle JDK, even on Ubuntu -- thanks for the tip. Please let me know if there are other suggestions.

File: addons/binding/org.openhab.binding.folding/README.md [hunk 17]
```diff
@@ -0,0 +1,30 @@
+# Folding@home binding
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/README.md]: change to "Binding"

Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/README.md]: change to "Binding"

File: addons/binding/org.openhab.binding.folding/README.md [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+This binding can control multiple Folding@home clients and slots, using the TCP
+interface. Clients are modeled as bridges, and support adding any number of slots
+(though, usually CPU and GPU). The binding provides control over Run / Pause and
+Finish. It polls for the status of the client, updates the run / pause state, and
+provides a basic description of the slot. 
+
+The clients must be added manually in the Paper UI, but the slots for that 
+client will then appear using auto-discovery.
+
+## Requirements (network access to F@H)
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/README.md]: add an empty line below every header

File: addons/binding/org.openhab.binding.folding/ESH-INF/thing/thing-types.xml [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+        <channels>
+            <channel id="run" typeId="run"/>
+            <channel id="finish" typeId="finish"/>
+        </channels>
+            
+        <config-description>
+            <parameter name="host" type="text">
+                <label>Host</label>
+                <description>Hostname or IP address</description>
+                <default>127.0.0.1</default>
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/ESH-INF/thing/thing-types.xml]: please also add
```
<context>network-address</context>
```

File: addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/discovery/FoldingDiscoveryProxy.java [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+ */
+package org.openhab.binding.folding.discovery;
+
+import org.eclipse.smarthome.core.thing.ThingUID;
+
+/**
+ * Singleton instance to connect the Client handler and the discovery
+ * service.
+ */
+
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/discovery/FoldingDiscoveryProxy.java]: remove this empty line and add an author tag to the javadoc

File: addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/discovery/FoldingSlotDiscoveryService.java [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.folding.FoldingBindingConstants;
+
+public class FoldingSlotDiscoveryService extends AbstractDiscoveryService {
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/discovery/FoldingSlotDiscoveryService.java]: please add a short JavaDoc here as well

File: addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/handler/SlotHandler.java [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SlotHandler extends BaseThingHandler implements SlotUpdateListener {
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/handler/SlotHandler.java]: please add a short JavaDoc

File: addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/handler/SlotInfo.java [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.folding.handler;
+
+import java.util.Map;
+
+public class SlotInfo {
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/handler/SlotInfo.java]: please add a short JavaDoc

File: addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/handler/SlotUpdateListener.java [hunk 24]
```diff
@@ -0,0 +1,15 @@
...
+ * Copyright (c) 2014-2016 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.folding.handler;
+
+public interface SlotUpdateListener {
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/handler/SlotUpdateListener.java]: please add a short JavaDoc

File: addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/internal/FoldingHandlerFactory.java [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.openhab.binding.folding.FoldingBindingConstants;
+import org.openhab.binding.folding.handler.FoldingClientHandler;
+import org.openhab.binding.folding.handler.SlotHandler;
+
+/**
+ * The {@link FoldingHandlerFactory} is responsible for creating things and thing
+ * handlers.
+ *
+ * @author FaMaKe - Initial contribution
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/src/main/java/org/openhab/binding/folding/internal/FoldingHandlerFactory.java]: please put your real name here

File: addons/binding/org.openhab.binding.folding/META-INF/MANIFEST.MF [hunk 26]
```diff
@@ -0,0 +1,27 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
+Bundle-Name: Folding Binding
+Bundle-SymbolicName: org.openhab.binding.folding;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.0.0.qualifier
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.folding/META-INF/MANIFEST.MF]: you need to change this to 2.1.0 as well

**Comment by fa2k** [comment]
Thanks for the review. I'll try to get these done more quickly than the last ones:)

**Comment by fa2k** [comment]
All right I think I have squashed it. Hope it looks OK.

**Pull request merged**